RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc

```diff
@@ -0,0 +1,117 @@
+
+    paramTS_ = std::make_unique<HcalRecoParams>(params);
+    paramTS_->setTopo(&htopo);
+
+    // Build the product
+    ReturnType prod(new HcalChannelPropertiesVec(htopo.ncells()));
+    std::array<HcalPipelinePedestalAndGain, 4> pedsAndGains;
+    const HcalSubdetector subdetectors[3] = {HcalBarrel, HcalEndcap, HcalForward};
+
+    for (HcalSubdetector subd : subdetectors) {
+      const HcalGeometry* hcalGeom = static_cast<const HcalGeometry*>(geom.getSubdetectorGeometry(DetId::Hcal, subd));
+      const std::vector<DetId>& ids = hcalGeom->getValidDetIds(DetId::Hcal, subd);
+
+      for (const auto cell : ids) {
+        const auto rawId = cell.rawId();
+
+        // ADC decoding tools, etc
+        const HcalRecoParam* param_ts = paramTS_->getValues(rawId);
+        const HcalQIECoder* channelCoder = cond.getHcalCoder(cell);
+        const HcalQIEShape* shape = cond.getHcalShape(channelCoder);
+        const HcalSiPMParameter* siPMParameter = cond.getHcalSiPMParameter(cell);
+
+        // Pedestals and gains
+        const HcalCalibrations& calib = cond.getHcalCalibrations(cell);
+        const HcalCalibrationWidths& calibWidth = cond.getHcalCalibrationWidths(cell);
+        for (int capid = 0; capid < 4; ++capid) {
+          pedsAndGains[capid] = HcalPipelinePedestalAndGain(calib.pedestal(capid),
+                                                            calibWidth.pedestal(capid),
+                                                            calib.effpedestal(capid),
+                                                            calibWidth.effpedestal(capid),
+                                                            calib.respcorrgain(capid),
+                                                            calibWidth.gain(capid));
+        }
+
+        // Channel quality
+        const HcalChannelStatus* digistatus = qual.getValues(rawId);
+        const bool taggedBadByDb = severity.dropChannel(digistatus->getValue());
+
+        // Fill the table entry
+        const unsigned linearId = htopo.detId2denseId(cell);
+        prod->at(linearId) =
+            HcalChannelProperties(&calib, param_ts, channelCoder, shape, siPMParameter, pedsAndGains, taggedBadByDb);
+      }
+    }
+
+    return prod;
+  }
+
+  HcalChannelPropertiesEP() = delete;
+  HcalChannelPropertiesEP(const HcalChannelPropertiesEP&) = delete;
+  HcalChannelPropertiesEP& operator=(const HcalChannelPropertiesEP&) = delete;
+
+private:
+  edm::ESGetToken<HcalDbService, HcalDbRecord> condToken_;
+  edm::ESGetToken<HcalTopology, HcalRecNumberingRecord> topoToken_;
+  edm::ESGetToken<HcalRecoParams, HcalRecoParamsRcd> paramsToken_;
+  edm::ESGetToken<HcalSeverityLevelComputer, HcalSeverityLevelComputerRcd> sevToken_;
+  edm::ESGetToken<HcalChannelQuality, HcalChannelQualityRcd> qualToken_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> geomToken_;
+
+  std::unique_ptr<HcalRecoParams> paramTS_;
```

Comment:
An ESProducer member data that gets changed for each IOV and is referenced by the produced ESProducts is incompatible with concurrent IOVs.

If I understood correctly, a pointer to the `HcalRecoParams` is passed to each `HcalChannelProperties` that need to keep the pointer as member data. Therefore concurrent IOVs could be supported if the ownership of the `HcalRecoParams` object would be given to the EventSetup system. I can think of three options
* modify `HcalChannelPropertiesVec` to hold the `HcalRecoParams` as well
  * would imply that `HcalChannelPropertiesVec` could not be simply `std::vector<HcalChannelProperties>` but a custom class with enough `vector`-like interface
* make `paramTS_` a `std::shared_ptr` and give each `HcalChannelProperties` a shared ownership to `HcalRecoParams`
  * sounds a bit overkill given that all `HcalChannelProperties` have the same life time
* split the `HcalChannelPropertiesEP::produce()` into two
  * first one produces the `paramTS_` into the `HcalChannelPropertiesRecord`
  * second one reads the `paramTS_` from `HcalChannelPropertiesRecord` and produces the `HcalChannelPropertiesVec`
     * `HcalChannelProperties` can continue use a raw pointer to refer to the `HcalRecoParams`, proper object life times are guaranteed by the EventSetup system

To me the last option sounds the cleanest one.

Reply:
The first option looks simplest to me. But could you explain why "an ESProducer member data that gets changed for each IOV and is referenced by the produced ESProducts is incompatible with concurrent IOVs"? The product is allocated on the heap, same way as a new version of HcalRecoParams. So whatever memory model is used to synchronize product making should work in this case as well.

Reply:
OK, never mind. I thought about it a little more and understood what the problem is.

Reply:
In your third option, do I have to create another record? If I don't, I get a compilation error which says "Trying to get a product with an ESGetToken specifying a Record from another Record where the second Record is not dependent on the first Record.
         static_assert(!std::is_same<FoundItrT, EndItrT>::value,"

Can I have HcalChannelPropertiesRecord depend on itself?

Reply:
> In your third option, do I have to create another record?

Same record should be fine.

Could you share the `ESGetToken` type and the line that gets the `HcalRecoParams` from  `HcalChannelPropertiesRecord`?

Reply:
OK, I resolved the dependency issue by introducing another record

Reply:
I tried what you suggested before, but I was getting the compilation error above. The system thinks that I can not create that token because HcalChannelPropertiesRecord does not depend on HcalChannelPropertiesRecord

Reply:
I took a look out of curiosity, and
https://github.com/makortel/cmssw/commit/536c28537f6ac388cbf2951dd7f4b950dc09678b
should be sufficient to get rid of the `HcalChannelPropertiesAuxRecord`. The issue is that if we have dependent records like `A -> B -> C` (`A` depends on `B` that depends on `C`), an `ESProducer::produce(A const& rcd)` can
* get another product from `A` with `rcd.get(tokenA)`
* get a product from `B` also with `rcd.get(tokenB)`
* get a product from `C` with `rcd.getRecord<B>.get(tokenC)`

i.e. the `rcd.get(token)` extends only to the immediate parents of `rcd`. It's a known deficiency, but so far has not felt to be a big-enough problem to justify the development cost.

Some of the other `get()`s can also be simplified
https://github.com/makortel/cmssw/commit/090f84ef2ac5735dee2381fe3a9a96d62e40da13

I leave it to your and reviewer's discretion whether to simplify or not.

Reply:
Well, it seems to work as it is. Another record solves the dependency issue in the most transparent manner (at least to me).